text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+(Following the principal of macro hygiene)
+Use of an unbound variable in the body is a syntax error.
+A parameter list may not declare the same variable twice, and may not list any standard Cedar variables.
+An unused variable is a syntax warning.
+Function and variable names share the same namesapce, with standard lexical scoping.
+Inside of a function, any function application (see below) that does not resolve to an extension function or built-in operation is an error. In other words, Cedar functions are not permitted to call other Cedar functions.
+
+### Function applications (a.k.a. function calls)
+
+A function application has the same syntax as an extension function constructor application. In particular, a function application is composed of two elements:
+
+1. The function name (potentially namespaced)
+2. A comma separated list of arguments
+
+Here is an example:
+
+```
+foo(1,2, principal.name)
+```
+
+Function arguments are eagerly evaluated (i.e., call-by-value), as with extension functions today.
+Functions do not exist at run-time -- they cannot be stored in entity attributes, requests, etc. As such, referencing a function by its name, without calling it, is a syntax error.
+Arguments for each of a function's declared parameters, and no more, must be provided at the call, or it's a syntax error.
+Other errors (such as type errors or overflow) are detected at runtime, or via validation.
+Examples:
+```
+function foo(a, b) {
+    a + b
+};
+
+permit(principal,action,resource) when {
+  foo + 1 // Parse error
+};
+permit(principal,action,resource) when {
+    foo(1) // Parse error
+};
+permit(principal,action,resource) when {
+    foo(1, 2) // No error
+};
+permit(principal,action,resource) when {
+    foo(1, "hello") // Run time type error
+};
+permit(principal,action,resource) when {
+    foo(1, "hello", principal) // Parse error
+};
+permit(principal,action,resource) when {
+    bar(1, "hello", principal) // Parse error
+};
+```
+
+### Namespacing/scoping
+
+All functions in a policyset are in scope for all policies, i.e., function declarations do not need to lexically precede use.
+Cedar policies and function bodies are lexically scoped.
+`principal`/`action`/`resource`/ `context` are considered to be more tightly bound then function names.
+This means that while you could name a function `principal`, you could never call it. (This should probably be a validator warning)
+Function name conflicts at the top level are a parse error.
+Function names may shadow extension functions (results in a warning).
+
+### Formal semantics
+This RFC adds one new evaluation rule to Cedar:
+If $f$ is the name of a declared function, _def_($f$) is the body of the definition, and $p_1, ..., p_n$ is the list of parameters in the definition:
+$f(v_1, ..., v_n) \rightarrow$ _def_$(f) [p_1 \mapsto v_1, ..., p_n \mapsto v_n]$
+Where $e[x \mapsto v]$ means to substitute the $v$ for $x$ in $e$, as usual.
```

Comment:
Per my comment above, I'd propose to do lazy evaluation instead. In other words:
```
def f(x1,x2) { body }
permit(...) when { ... f(e1,e2) ... };
```
is equivalent to
```
permit(...) when { ... body[x1 --> e1, x2 --> e2] ... }
```